### PR TITLE
UCS/BUILD: Add infrastructure for loadable modules

### PR DESCRIPTION
**Pull request by yosefe**
Adds the UCS service for dlopen()-ing shared libraries, as preparation for moving some ucm/uct components which have external deps to separate shared libraries

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8457/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8458/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8459/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8460/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8474/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8478/ for details (Mellanox internal link).

File: src/ucs/sys/module.c [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        return;
+    }
+
+    ucs_debug("ucs library path: %s", dl_info.dli_fname);
+
+    /* copy extension */
+    dlpath_dup = ucs_strdup(dl_info.dli_fname,
+                            UCS_MODULE_PATH_MEMTRACK_NAME);
+    if (dlpath_dup == NULL) {
+        goto out;
```
Comment (brminich) [inline src/ucs/sys/module.c]: no need to `goto out`, because `dlpath_dup` was not allocated, can just return here

File: src/ucs/sys/module.c [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        strncpy(ucs_module_loader_state.module_ext, p,
+                sizeof(ucs_module_loader_state.module_ext) - 1);
+    }
+    ucs_free(dlpath_dup);
+
+    /* copy directory component */
+    dlpath_dup = ucs_strdup(dl_info.dli_fname,
+                            UCS_MODULE_PATH_MEMTRACK_NAME);
+    if (dlpath_dup == NULL) {
+        goto out;
```
Comment (brminich) [inline src/ucs/sys/module.c]: need to return

File: config/module.am [hunk 3]
```diff
@@ -0,0 +1,26 @@
+#
+# Copyright (C) Mellanox Technologies Ltd. 2001-2018.  ALL RIGHTS RESERVED.
```
Comment (hoopoepg) [inline config/module.am]: 2019?

File: src/ucs/sys/module.c [hunk 4]
```diff
@@ -0,0 +1,164 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2001-2018.  ALL RIGHTS RESERVED.
```
Comment (hoopoepg) [inline src/ucs/sys/module.c]: 2019?

File: config/module.am (hunk 3 above)
Comment (yosefe) [reply config/module.am]: well, i've added this code on 2018 :)

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8520/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8525/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8528/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8537/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8540/ for details (Mellanox internal link).

**Pull request merged**